### DISCUSSION TITLE
How to pass arguments to decorators?

### DISCUSSION THREAD

**Discussion by rcarmo** [hylang/hy#2500] [Q&A] [answered]
I need to use a set of decorators that take arguments. I've tried this:

```hylang
(defn [(partial decorator "foo")] bar ... }
```

But it doesn't seem to work. Any ideas? I could do this with `with-decorator` and the old `#@` macro in past versions, but the constant syntax changes make it hard to maintain Hy code...

**Comment by Kodiologist** [comment]
You just call the function with arguments, same as before, and as in Python. 

> But it doesn't seem to work

Can you be more specific?

**Comment by rcarmo** [comment]
I _think_ I've finally got it, but it seems overly contrived:

```hylang
(defn tracer [func #* args]
    (fn [#* args] 
        (do (print #* args) 
            (func #* args))))

(defn [tracer] test [x]
  (print x)

(import functools [partial])

(defn tracer2 [prefix func #* args]
  (fn [#* args]
    (do (print prefix #* args) 
        (func #* args))))

(defn [(partial tracer2 ">")] test2 [x]
  (print x))
```

This works, but I've scoured the documentation and tests and they do not provide good enough hints as to the best way to create and use decorators with arguments in "modern" Hy, so any pointers would be welcome (and yes, I need to both use arguments for existing decorators and export others out to Python).

> As an aside, I've been patiently poking at https://github.com/rcarmo/sushy over the past week and after seven years there have been so many breaking changes in Hy that it's been a somewhat frustrating experience to refactor it to "almost 1.0" shape. But at least `let` and `async` work, so I can get stuff done...

**Reply by Kodiologist** [reply]
I'm not sure I understand what's contrived about it. This is the same way as how it works in Python. You just put the decorator code in the square brackets, instead of in front of the `def` with an at-sign. Is there some way in which the translation from Python to Hy is unclear?

**Reply by wrobell** [reply]
I think the confusion here is around `partial` and order of arguments. Typical decorator takes function as first parameter. `partial(tracer, "foo")` replaces decorated function with "foo" string.

@rcarmo You probaly want to use a keyword argument with partial.

**Comment by rcarmo** [comment]
The order of arguments means interop is fiddly. I’m unwrapping older decorators I had written (that worked flawlessly either way) and coming across this time and time again.

(Also, I don’t see keyword arguments as a solution for this)

I can sort of make it work, but wish there was proper documentation (and written tests) for it that accounted for Python interop both ways.

**Reply by wrobell** [reply]
What is the equivalent in Python, which works for you?

**Reply by Kodiologist** [reply]
But the order of arguments is the same as in Python, and the current documentation on decorators appears to be complete (#2391), and we already have tests for decorators. I don't know what to tell you, man.

**Reply by wrobell** [reply]
> (Also, I don’t see keyword arguments as a solution for this)

If you are writing your own decorators, then I would not use partial as well. Good example of a decorator is `dataclass` one https://github.com/python/cpython/blob/main/Lib/dataclasses.py. You can use it with and without parameters.

I am not sure what originally worked for you with older Hylang version, but looking at your examples they would not work in Python as well (but maybe we are miscommunicating here).

**Comment by rcarmo** [comment]
Well, just to give an idea of what I'm currently doing, part of the code I'm "porting" back to Hy is stuff like [this decorator](https://github.com/rcarmo/sushy/blob/ce5272289e955736236f2448a2dd0b745275aaf7/sushy/routes.hy#L92), which was used in several HTTP routes like this:

```hylang
(with-decorator 
    (handle-get (+ *page-route-base* "/<pagename:path>"))
    (report-processing-time)
    (http-caching "pagename" "text/html" 3600)
    (ttl-cache 30)
    (render-view "wiki")
    (defn wiki-page [pagename] 
        (if (in (.lower pagename) *redirects*) ...
```

Due to the instability of earlier Hy releases, seven years ago I rewrote that in Python 3 like so (with minor changes since some of the tracing and caching had to allow for live runtime value updates):

```python
@get(PAGE_ROUTE_BASE + "/<pagename:path>")
@report_processing_time(STATS_ADDRESS, STATS_PORT)
@http_caching("pagename", "text/html", ASSET_HTTP_EXPIRY)
@lru_cache(10)
@view("wiki")
def wiki_page(pagename):
```

Right now I'm trying to, in parallel, get both [my public project](https://github.com/rcarmo/sushy) and other, similar services working on "modern" Hy, hence my looking for better examples. My non-public projects have to use similar decorators (some of them with async frameworks, which is a minor difference), but those have to be shared with Python code that I'm not going to rewrite.

**Reply by Kodiologist** [reply, marked as answer]
So you'd now write that as:


```
(defn
  [
   (get (+ PAGE_ROUTE_BASE "/<pagename:path>"))
   (report_processing_time STATS_ADDRESS STATS_PORT)
   (http_caching "pagename" "text/html" ASSET_HTTP_EXPIRY)
   (lru_cache 10)
   (view "wiki")]
  wiki_page [pagename]
  …)
```

except you should rename `get` or it will be shadowed by the core macro of the same name. Straightforward enough, no? Am I missing something?

**Reply by rcarmo** [reply]
Yes, I cannot get the decorators themselves to work properly. These are just examples of why I can't use keyword arguments, etc.

**Reply by Kodiologist** [reply]
That probably doesn't have to do with how you're calling the decorators on `wiki_page` or passing them arguments, then, but how you implemented the decorators.

**Reply by rcarmo** [reply]
Well, they used to work fine. But seven years of breaking change have obviously made things a trifle hard.
